Make save-cloud easier to contribute for mac users (#699)
analysis-dev/save-cloud: CONTRIBUTING.md

# Contributing
1. Fork this repository to your own account

> 2. Make your changes and verify that tests pass

Reviewer: Pls add specific commands on how to verify tests

Author: `save-cloud:check` task, isn't it?

Reviewer: No, it should be something like `./gradlew check` or even `./gradlew build`

Author: ok, will do it

3. Commit your work and push to a new branch on your fork
4. Submit a pull request
5. Participate in the code review process by responding to feedback

> ### Mac M1 contributors

Reviewer 2: please add an instruction that you need to install the image of mysql:
``

Author: There is no need in it because if we set `platform` flag in docker configuration, everything works fine on M1. Need to check on other platforms.

In order to launch the project locally, you need to do these preparations:

> 1. In file `save-cloud/build.gradle.kts` change languageVersion of `org.liquibase.gradle.LiquibaseTask` from 11 to 17

Reviewer: We shuold try and change version and see if build passes. Liquibase might have fixed issue with newer JVMs, because we observed failures with Liquibase 4.7, and now it's already 4.9

Author: Could you please try in on Windows? On Mac it works fine.

Reviewer: On windows I get
```
The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.
	at liquibase.database.ConnectionServiceFactory.create(ConnectionServiceFactory.java:35)
	at liquibase.database.DatabaseFactory.openConnection(DatabaseFactory.java:214)
	... 31 more
Caused by: liquibase.exception.DatabaseException: Connection could not be created to jdbc:mysql://localhost:3306/save_cloud with driver com.mysql.cj.jdbc.Driver.  Communications link failure
```
with JVM 17

Reviewer: I've found [this](https://stackoverflow.com/a/67438592) SO answer, and adding this flag to URL takes things somewhat further, but still ends with:
```
[2022-04-22 14:01:03] SEVERE [liquibase.integration] Cannot invoke "java.io.InputStream.available()" because "is" is null
liquibase.exception.CommandExecutionException: liquibase.exception.LiquibaseException: Unexpected error running Liquibase: Cannot invoke "java.io.InputStream.available()" because "is" is null
```

Author: oh ok thanks

Author: Made it to choose java version according to os family.

Reviewer: Funny thing, it actually works on Windows, if you *launch* gradle itself with JVM 16 or 17, i.e. `JAVA_HOME` points to a new Java. With `JAVA_HOME` being Java 8 or 11 or even 13 `liquibaseUpdate` fails with Java 17 in toolchains. Very strange behavior which actually looks like bug on the side of Gradle.

so there would be something like this:
```
tasks.withType<org.liquibase.gradle.LiquibaseTask>().configureEach {
    this.javaLauncher.set(project.extensions.getByType<JavaToolchainService>().launcherFor {
        languageVersion.set(JavaLanguageVersion.of(17))
    })
}
```

> 2. If you use Spring Idea Plugin on Mac, you should set `Active Profiles` to be `dev, mac, secure` for `SaveApplication`

Reviewer: ```suggestion
2. If you use Spring Idea Plugin (built into IDEA Ultimate) on Mac, you should set `Active Profiles` to be `dev, mac, secure` for `SaveApplication`
```

> and `mac` for 

Reviewer: for?..

Author: fixed

# Contributing
1. Fork this repository to your own account
2. Make your changes and verify that tests pass (with `check` gradle task)
3. Run `diktat` static analyzer with `diktatFix` gradle task or `diktatCheck` task depending on what you want.

> 4. Run `detekt` with `detekt` gradle task

Reviewer: I think, it's better to use `detektAll`, because AFAIR `detekt` task is only for JVM

Author: guess you are right, will fix it

5. After all tests and analyzers passing commit your work and push to a new branch on your fork
6. Submit a pull request
7. Participate in the code review process by responding to feedback

## Launching save-cloud with command line
1. Use `gradlew.bat startMysqlDb` on Windows and `./gradlew startMysqlDb` on other platforms for setting up database.
Make sure you have Docker installed and active.
2. Run backend.

> It can be run either with `./gradlew save-backend:bootRun` or with Intellij Idea Ultimate plugin called `Services`.

Reviewer: ```suggestion
It can be run either with `./gradlew save-backend:bootRun` or with Intellij Idea Ultimate plugin from the menu `Services`.
```

3. Run frontend. It can be run with `./gradlew save-frontend:run`.

> You can turn continuous mode on passing `--continuous` flag.

Reviewer: ```suggestion
You can enable hot reload by passing `--continuous` flag.
```

> ## Services Intellij Idea Ultimate plugin

Reviewer: `Services` is not a plugin, it's just a section of menu, which is integrated with Spring support plugin

Author: oh ok
